File: bilmarknad-mcp/src/bilmarknad_mcp/blocket_proxy.py
from __future__ import annotations

import os
from typing import Any

import httpx

from bilmarknad_mcp.blocket import parse_ad
from bilmarknad_mcp.schema import CarListing

DEFAULT_PROXY_BASE = "https://blocket-api.se"


def proxy_base_url() -> str:
    return (os.environ.get("BLOCKET_PROXY_URL") or DEFAULT_PROXY_BASE).rstrip("/")


class BlocketProxyClient:
    def __init__(self, client: httpx.Client | None = None, base_url: str | None = None):
        self._client = client
        self._owns = client is None
        self._base_url = (base_url or proxy_base_url()).rstrip("/")

    def _get_client(self) -> httpx.Client:
        if self._client is None:
            self._client = httpx.Client(timeout=30.0, headers={"User-Agent": "bilmarknad-mcp/0.1"})
        return self._client

    def search(self, **kwargs: Any) -> list[CarListing]:
        params = dict(kwargs)
        client = self._get_client()
        response = client.get(f"{self._base_url}/search", params=params)
        if response.status_code >= 400:
            return []
        payload = response.json()
        docs = payload.get("docs") or payload.get("results") or payload
        if not isinstance(docs, list):
            return []
        return [parse_ad(item) for item in docs]

    def close(self) -> None:
        if self._owns and self._client is not None:
            self._client.close()

File: internal/bilmarknad/blocketproxy/blocketproxy.go
package blocketproxy

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"

	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/blocket"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/httputil"
	"github.com/flyhard/swedish-car-mcp/internal/bilmarknad/schema"
)

const defaultProxyBase = "https://blocket-api.se"

// ProxyBaseURL returns the configured Blocket proxy base URL.
func ProxyBaseURL() string {
	if v := os.Getenv("BLOCKET_PROXY_URL"); v != "" {
		return strings.TrimRight(v, "/")
	}
	return defaultProxyBase
}

// Client queries Blocket via an optional proxy service.
type Client struct {
	httpClient *http.Client
	owns       bool
	baseURL    string
}

func NewClient(c *http.Client, baseURL string) *Client {
	if c == nil {
		c = httputil.NewClient()
	}
	if baseURL == "" {
		baseURL = ProxyBaseURL()
	}
	return &Client{httpClient: c, owns: c.Transport == nil, baseURL: strings.TrimRight(baseURL, "/")}
}

func (c *Client) Search(ctx context.Context, params map[string]string) ([]schema.CarListing, error) {
	u := c.baseURL + "/search"
	reqURL, _ := url.Parse(u)
	q := reqURL.Query()
	for k, v := range params {
		q.Set(k, v)
	}
	reqURL.RawQuery = q.Encode()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqURL.String(), nil)
	if err != nil {
		return nil, err
	}
	resp, err := httputil.DoWithRetry(ctx, c.httpClient, req, "blocket", httputil.DefaultRetryPolicy())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		return nil, nil
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var payload map[string]any
	if err := json.Unmarshal(body, &payload); err != nil {
		return nil, err
	}
	docs, _ := payload["docs"].([]any)
	if docs == nil {
		if results, ok := payload["results"].([]any); ok {
			docs = results
		} else if arr, ok := payload[""].([]any); ok {
			docs = arr
		}
	}
	if docs == nil {
		if arr, ok := any(payload).([]any); ok {
			docs = arr
		}
	}
	if docs == nil {
		return nil, nil
	}
	out := make([]schema.CarListing, 0, len(docs))
	for _, doc := range docs {
		if m, ok := doc.(map[string]any); ok {
			out = append(out, blocket.ParseAd(m))
		}
	}
	return out, nil
}

func (c *Client) Close() {
	if c.owns && c.httpClient != nil {
		c.httpClient.CloseIdleConnections()
	}
}
